packages/discovery-provider/src/tasks/index_payment_router.py

```diff
@@ -0,0 +1,858 @@
+        )
+        return False
+    return True
+
+
+class ReceiverUserAccount(TypedDict):
+    user_id: int
+    user_bank_account: str
+
+
+class PurchaseMetadataDict(TypedDict):
+    price: int
+    splits: dict[str, int]
+    type: PurchaseType
+    id: int
+    purchaser_user_id: int
+    content_owner_id: int
+
+
+# Return highest payment router slot that has been processed
+def get_highest_payment_router_tx_slot(session: Session):
+    slot = MIN_SLOT
+    tx_query = (
+        session.query(PaymentRouterTx.slot).order_by(desc(PaymentRouterTx.slot))
+    ).first()
+    if tx_query:
+        slot = tx_query[0]
+    return slot
+
+
+# Cache the latest value committed to DB in redis
+# Used for quick retrieval in health check
+def cache_latest_sol_payment_router_db_tx(redis: Redis, tx):
+    cache_latest_sol_db_tx(redis, latest_sol_payment_router_db_tx_key, tx)
+
+
+# Query a tx signature and confirm its existence
+def get_tx_in_db(session: Session, tx_sig: str) -> bool:
+    exists = False
+    tx_sig_db_count = (
+        session.query(PaymentRouterTx).filter(PaymentRouterTx.signature == tx_sig)
+    ).count()
+    exists = tx_sig_db_count > 0
+    return exists
+
+
+def get_purchase_metadata_from_memo(
+    session: Session, memos: List[str], timestamp: datetime
+) -> Union[PurchaseMetadataDict, None]:
+    """Checks the list of memos for one matching the format of a purchase's content_metadata, and then uses that content_metadata to find the premium_conditions associated with that content to get the price"""
+    for memo in memos:
+        try:
+            content_metadata = memo.split(":")
+            if len(content_metadata) == 4:
+                (
+                    type_str,
+                    id_str,
+                    blocknumber_str,
+                    purchaser_user_id_str,
+                ) = content_metadata
+                type = PurchaseType[type_str.lower()]
+                id = int(id_str)
+                purchaser_user_id = int(purchaser_user_id_str)
+                blocknumber = int(blocknumber_str)
+
+                # TODO: Wait for blocknumber to be indexed by ACDC
+                logger.debug(
+                    f"index_payment_router.py | Found content_metadata in memo: type={type}, id={id}, blocknumber={blocknumber} user_id={purchaser_user_id}"
+                )
+                price = None
+                splits = None
+                content_owner_id = None
+                if type == PurchaseType.track:
+                    env = shared_config["discprov"]["env"]
+                    content_owner_id = get_track_owner_id(session, id)
+                    if content_owner_id is None:
+                        logger.error(
+                            f"index_payment_router.py | Couldn't find content owner for track_id={id}"
+                        )
+                        continue
+                    query = session.query(TrackPriceHistory)
+                    if env != "dev":
```

Comment:
hopefully solder's bankrun replacement for solana-test-validator helps here eventually in dev
https://github.com/kevinheavey/solders